src/libostree/ostree-repo-pull.c

```diff
@@ -1322,6 +1324,165 @@ commitstate_is_partial (OtPullData   *pull_data,
 }
 
 static gboolean
+ensure_collection_ref_in_summary (GVariant   *summary,
+                                  const char *collection_id,
+                                  const char *ref)
+{
+  g_assert_nonnull (collection_id);
+  g_assert_nonnull (ref);
+
+  g_autoptr(GVariant) additional_metadata = NULL;
+  additional_metadata = g_variant_get_child_value (summary, 1);
+
+  const char *main_collection_id;
+  if (!g_variant_lookup (additional_metadata,
+                         OSTREE_SUMMARY_COLLECTION_ID,
+                         "&s",
+                         &main_collection_id))
+    main_collection_id = NULL;
+
+  gboolean check_main_collection;
+  check_main_collection = (g_strcmp0 (main_collection_id, collection_id) == 0);
+
+  if (check_main_collection)
+    {
+      g_autoptr(GVariant) refs = NULL;
+      gsize i, n;
+
+      refs = g_variant_get_child_value (summary, 0);
+      for (i = 0, n = g_variant_n_children (refs); i < n; i++)
```

Comment:
This can be optimised using a binary search, since the refs are guaranteed to be in alphabetical order. Use `ot_variant_bsearch_str()`.